vm/natives.c

```diff
@@ -1138,5 +1138,393 @@ const AsebaNativeFunctionDescription AsebaNativeDescription_rand =
+	if (index_val < dq_size / 2)
+	{
+		deque_shift(vm, dest, dq_capacity,
+					/* from */ dq_start + index_val + len_val - 1,
+					/* to   */ dq_start + len_val,
+					-len_val);
+		vm->variables[dest + 1] = dq_start = (dq_start + len_val) % dq_capacity;
+	}
+	// else in right half, shift suffix elements left
+	else
+	{
+		deque_shift(vm, dest, dq_capacity,
+					/* from */ dq_start + index_val,
+					/* to   */ dq_start + dq_size - 1 - len_val,
+					len_val);
+	}
+	vm->variables[dest] = dq_size = dq_size - len_val;
+}
+
+void AsebaNative_deqerase(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 index = AsebaNativePopArg(vm);
+	uint16 len = AsebaNativePopArg(vm);
+	uint16 index_val = vm->variables[index];
+	uint16 len_val = vm->variables[len];
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+	
+	// Check for exception
+	if (vm->variables[index] < 0 || index_val > dest_length - 2 - 1
+		|| len_val > vm->variables[dest] - index_val)
+	{
+		vm->flags = ASEBA_VM_STEP_BY_STEP_MASK;
+		AsebaSendMessage(vm, ASEBA_MESSAGE_ARRAY_ACCESS_OUT_OF_BOUNDS, &(vm->pc), sizeof(vm->pc));
+		return;
+	}
+	
+	// Do it
+	_AsebaNative_deqerase(vm, dest, index_val, len_val, dest_length);
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqerase =
+{
+	"deque.erase",
+	"shift dest at index by len to erase",
+	{
+		{ -1, "dest" },
+		{  1, "index" },
+		{  1, "len" },
+		{ 0, 0 }
+	}
+};
+
+void AsebaNative_deqpushfront(AsebaVMState *vm)
+{
+	// Pop arguments off
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+
+	// Now call insert with proper arguments
+	_AsebaNative_deqinsert(vm, dest, src, 0, dest_length, src_length);
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqpushfront =
+{
+	"deque.push_front",
+	"insert src before the front of dest",
+	{
+		{ -1, "dest" },
+		{ -2, "src" },
+		{ 0, 0 }
+	}
+};
+
+void AsebaNative_deqpushback(AsebaVMState *vm)
+{
+	// Pop arguments off
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+	
+	// Now call insert with proper arguments
+	_AsebaNative_deqinsert(vm, dest, src, vm->variables[dest], dest_length, src_length);
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqpushback =
+{
+	"deque.push_back",
+	"insert src after the back of dest",
+	{
+		{ -1, "dest" },
+		{ -2, "src" },
+		{ 0, 0 }
+	}
+};
+
+void AsebaNative_deqpopfront(AsebaVMState *vm)
+{
+	// Pop arguments off
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+
+	// call insert and erase with proper arguments
+	_AsebaNative_deqget(vm, dest, src, 0, dest_length, src_length);
```

Comment:
I would add a check for underflow for pop_front/back, as these are the high-level API.